### PR TITLE
Updated Documentation for the CI with mermaid sequence diagrams

### PR DESCRIPTION
**Pull request by potiuk**
Part of #10368

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
This is purely a documentation change describeing the future CI architecture as well adding mermaid support for generation of sequence diagrams so that it is easier to review.

**Comment by potiuk** [comment]
Ad discussed in #10368 - I am going to merge that one after the code is reviewed/fixed/merged

**Comment by potiuk** [comment]
Rebased after merging #10368 - I think it is ready for review and very accurately describes what we have in CI now.

**Comment by potiuk** [comment]
Would love to merge the docs to close the task!

File: CI.rst [hunk 1]
```diff
@@ -17,6 +17,10 @@
 
 .. contents:: :local:
 
+NOTE! This is the target architecture we are aspiring to. We have a
+`<work in progress PR https://github.com/apache/airflow/pull/10368>`_ which implements this
+architecture but the documentation generation and plan is merged separately as "future" change.
+
```
Comment (feluelle) [inline CI.rst]: This is not up to date anymore, is it?

File: CI.rst [hunk 2]
```diff
@@ -188,32 +578,104 @@ The issues that gets updated are different for different branches:
...
+.. code-block:: bash
+
+  docker pull docker.pkg.github.com/apache/airflow/master-python3.6-ci:210056909
+
+  docker run -it docker.pkg.github.com/apache/airflow/master-python3.6-ci:210056909
+
+
+But you usually need to pass more variables amd complex setup if you want to connect to a database or
+enable some integrations. Therefore it is easiest to use `Breeze <BREEZE.rs>`_ for that. For example if
```
Comment (feluelle) [inline CI.rst]:
```suggestion
enable some integrations. Therefore it is easiest to use `Breeze <BREEZE.rst>`_ for that. For example if
```

File: images/ci/pull_request_ci_flow.mermaid [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+
+sequenceDiagram
+    Contributor-->>Fork: Submit PR
```
Comment (feluelle) [inline images/ci/pull_request_ci_flow.mermaid]:
```suggestion
    Contributor -->> Fork: Submit PR
```

File: CI.rst (hunk 1 above)
Comment (potiuk) [reply CI.rst]: Correct @feluelle ! Thanks for looking into it :)

**Comment by kaxil** [comment]
Yes, I feel we could reduce the code to maintain by using something other than Mermaid, this is where I think draw.io might  be helpful

**Comment by kaxil** [comment]
https://drawio-app.com/create-mermaid-diagrams-in-draw-io/ -> There is an integration that is already available

**Comment by potiuk** [comment]
> https://drawio-app.com/create-mermaid-diagrams-in-draw-io/ -> There is an integration that is already available

So you basically say you want to still keep the sources of mermaid but rather than doing `git commit` and have them automatically generated you want me to always go to draw,io, copy&paste the code, generate the image, download them, replace the images, commit (and not forget to copy back the sources if I changed them in draw,io?) 

Because that's the alternative. Currently the only thing you need to do is to commit the sources and the images will be generated autinatically

This is like forcing me to do those manual steps where simple (literally 20 lines of integration code that will never change) can do it for me. I do not really see the logic behind it.

**Comment by potiuk** [comment]
I already plan some changes in those diragras - for example when I implement #10507 -  there will be quite some changes there and simply changing that in text is so much easier than having to do all those manual steps.

**Comment by kaxil** [comment]
I am definitely against have a CI-CD for a single tool, which if it breaks for whatever reason is a code debt we should not have.

These can be easily replaced by a generated PNGs.

**Comment by kaxil** [comment]
You don't see the logic behind it because you are not reasoning for the amount of code that needs to be maintained

**Comment by potiuk** [comment]
> I am definitely against have a CI-CD for a single tool, which if it breaks for whatever reason is a code debt we should not have.

It only runs in CI-CD if the source mermaid files change. We keep hashes of the files and no mermaid will be run when then .mermaid files are not touched.

**Comment by kaxil** [comment]
Maintaining a separate node file just for mermaid makes 0 sense

**Comment by potiuk** [comment]
> Maintaining a separate node file just for mermaid makes 0 sense

If package.json is the problem, then no worries.  I can remove the package.json and move it to the script generating mermaid (which will only be run when .mermaid files change). Woudl that work?

**Comment by kaxil** [comment]
Are the package.json and the files for building mermaid diagrams in some other PR? I remember seeing them

**Comment by potiuk** [comment]
> Are the package.json and the files for building mermaid diagrams in some other PR? I remember seeing them

Just looking at that. I think they are lost on rebase :)

**Comment by potiuk** [comment]
> > Are the package.json and the files for building mermaid diagrams in some other PR? I remember seeing them
> 
> Just looking at that. I think they are lost on rebase :)

I will restore them in the way I think they can work without package.json.

**Comment by potiuk** [comment]
It's there. Standalone bash `pre-commit` script that will only do anything if any of the .mermaid files changed vs. hashes stored next to them. 

That "anything" is installing mermaid in a separate node virtualenv (in ,build/mermaid  folder - the ".build" folder is already all but ignored due to being used by breeze). It will install mermaid there, first time when it is run and keep on reusing it every next time it is run. 

If there is no change in those files or if you have no npm installed, the pre-commit will silently succeed.  If you do not touch any of those files, it's a no-op - both for the developers and for the CI. First of all pre-commit will be skipped if those files are not changed, secondly - even if you run pre-commit with `--all-files` those files will be skipped as the .md5 hashes will not change for them. And even if they do, but you have npm, they will be silently ignored.

I even made sure not to commit the mermaid-config.json - it is dynamically generated by the self-contained bash script. I expect this script to never change (unless we want to update colors/fonts etc. in the generated mermaid files).

**Review by kaxil — APPROVED** [review]

**Comment by kaxil** [comment]
CI is failing, maybe needs a rebase

**Pull request merged**